tools/metrics/metrics.go

```diff
@@ -0,0 +1,131 @@
+package main
+
+import (
+	"context"
+	"log"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	Gauge   eventKind = "gauge"
+	Counter eventKind = "counter"
+)
+
+var (
+	EOS = errors.New("End Of Stream")
+)
+
+type eventKind string
+
+type MetricWriter interface {
+	WriteMetrics([]Metric) error
+}
+type MetricReader interface {
+	ReadMetric() (Metric, error)
+}
+
+type MetricCloser interface {
+	Close()
+}
+
+type MetricReadCloser interface {
+	MetricReader
+	MetricCloser
+}
+
+type Metric struct {
+	ID    string
+	Kind  eventKind
+	Name  string
+	Time  time.Time
+	Value float64
+}
+
+type PollFunc func(MetricWriter) error
+
+func NewMetricPoller(interval time.Duration, fn PollFunc) MetricReadCloser {
+	buf := NewMetricBuffer(8) // 8 = buffer size ... arbitarily chosen
+	go func() {
+		for {
+			select {
+			case <-buf.ctx.Done():
+				return
+			default:
+				err := fn(buf)
+				if err != nil {
+					log.Println(err, "...continuing") // THINK: should probably close and exit the Reader to let errors bubble up
+				}
+			}
+			time.Sleep(interval)
+		}
+	}()
+	return buf
+}
+
+type MetricBuffer struct {
+	events chan Metric
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func (eq *MetricBuffer) WriteMetrics(events []Metric) error {
+	for _, ev := range events {
+		eq.events <- ev
+	}
+	return nil
+}
+
+func (eq *MetricBuffer) ReadMetric() (Metric, error) {
+	select {
+	case <-eq.ctx.Done():
+		return Metric{}, EOS
+	default:
+		return <-eq.events, nil
+	}
+}
+
+func (eq *MetricBuffer) Close() {
+	eq.cancel()
+}
+
+func NewMetricBuffer(n int) *MetricBuffer {
+	events := make(chan Metric, n)
+	ctx, cancel := context.WithCancel(context.Background())
+	return &MetricBuffer{events, ctx, cancel}
+}
+
+func NewMultiMetricReader(rs ...MetricReader) MetricReadCloser {
+	if len(rs) < 1 {
+		panic("you need at least one reader")
```

Comment:
That check isn't really needed tbh ... so yeah I'll just remove it.

I like the Readers all being composable and the errors coming out during ReadEvent. :)